java/src/test/java/org/apache/arrow/gandiva/evaluator/NativeEvaluatorTest.java

```diff
@@ -235,6 +265,102 @@ public void testAdd3() throws GandivaException, Exception {
     eval.close();
   }
 
+  @Test
+  public void testStringFields() throws GandivaException {
+    /*
+     * when x < "hello" then octet_length(x)
+     * else bit_length(x)
+     */
+
+    Field x = Field.nullable("a", new ArrowType.Utf8());
+    ArrowType retType = new ArrowType.Int(32, true);
+
+    TreeNode cond = TreeBuilder.makeFunction("less_than",
+      Lists.newArrayList(TreeBuilder.makeField(x), TreeBuilder.makeStringLiteral("hello")),
+      boolType);
+    TreeNode octetLenFuncNode = TreeBuilder.makeFunction("octet_length",
+      Lists.newArrayList(TreeBuilder.makeField(x)),
+      retType);
+    TreeNode bitLenFuncNode = TreeBuilder.makeFunction("bit_length",
+      Lists.newArrayList(TreeBuilder.makeField(x)),
+      retType);
+
+    TreeNode ifHello = TreeBuilder.makeIf(cond, octetLenFuncNode, bitLenFuncNode, retType);
+
+    ExpressionTree expr = TreeBuilder.makeExpression(ifHello, Field.nullable("res", retType));
+    Schema schema = new Schema(Lists.newArrayList(x));
+    NativeEvaluator eval = NativeEvaluator.makeProjector(schema, Lists.newArrayList(expr));
+
+    int numRows = 4;
+    byte[] validity = new byte[]{(byte) 255, 0};
+    String[] valuesX = new String[]{"hell", "abc", "hellox", "ijk" };
+    int[] expected = new int[]{4, 3, 48, 24};
+
+    ArrowBuf validitya = buf(validity);
+    List<ArrowBuf> inBufsA = stringBufs(valuesX);
+
+    ArrowRecordBatch batch = new ArrowRecordBatch(
+      numRows,
+      Lists.newArrayList(new ArrowFieldNode(numRows, 0), new ArrowFieldNode(numRows, 0)),
+      Lists.newArrayList(validitya, inBufsA.get(0), inBufsA.get(1)));
+
+    IntVector intVector = new IntVector(EMPTY_SCHEMA_PATH, allocator);
+    intVector.allocateNew(numRows);
+
+    List<ValueVector> output = new ArrayList<ValueVector>();
+    output.add(intVector);
+    eval.evaluate(batch, output);
+
+    for (int i = 0; i < numRows; i++) {
+      assertFalse(intVector.isNull(i));
+      assertEquals(expected[i], intVector.get(i));
+    }
+    eval.close();
+  }
+
+  @Test
+  public void testBinaryFields() throws GandivaException {
+    Field a = Field.nullable("a", new ArrowType.Binary());
+    Field b = Field.nullable("b", new ArrowType.Binary());
+    List<Field> args = Lists.newArrayList(a, b);
+
+    ArrowType retType = new ArrowType.Bool();
+    ExpressionTree expr = TreeBuilder.makeExpression("equal", args,
+      Field.nullable("res", retType));
+
+    Schema schema = new Schema(Lists.newArrayList(args));
+    NativeEvaluator eval = NativeEvaluator.makeProjector(schema, Lists.newArrayList(expr));
+
+    int numRows = 4;
+    byte[] validity = new byte[]{(byte) 255, 0};
+    String[] valuesA = new String[]{"a", "aa", "aaa", "aaaa"};
```

Comment:
should we also try with a non ascii character - to ensure things are ok?

Reply:
will do.

Reply:
Can we also do a combination of varlen input and fixedlen input?

if (x < "hello") y else z

Reply:
Do you mean y & z to be string fields ? we don't support that (var-ten output fields yet).

The testStringFields test is similar (if x < "hello" octet_length(x) else bit_length(x))